api/queue/worker.go

```diff
@@ -0,0 +1,84 @@
+package queue
+
+import (
+	"time"
+
+	"github.com/gojek/merlin/log"
+	"github.com/jinzhu/gorm"
+)
+
+var (
+	findJobQuery           = "SELECT * FROM jobs where id = ? AND completed = ? FOR UPDATE SKIP LOCKED"
+	allIncompleteJobsQuery = "SELECT * FROM jobs where completed = ? FOR UPDATE SKIP LOCKED"
+)
+
+type worker struct {
+	quitChan   chan bool
+	jobFuncMap map[string]JobFn
+	jobChan    chan *Job
+	db         *gorm.DB
+}
+
+func newWorker(db *gorm.DB, jobChan chan *Job) *worker {
+	quitChan := make(chan bool)
+	return &worker{db: db, jobChan: jobChan, quitChan: quitChan}
+}
+
+func (w *worker) updateWorkerJobFunction(jobFn map[string]JobFn) {
+	w.jobFuncMap = jobFn
+}
+
+func (w *worker) start() {
+	go func() {
+		for {
+			select {
+			case <-w.quitChan:
+				return
+			case job := <-w.jobChan:
+				go w.processJob(job)
+			}
+		}
+	}()
+}
+
+func (w *worker) processJob(job *Job) {
+	var refreshedJob Job
+	tx := w.db.Begin()
+	err := tx.Raw(findJobQuery, job.ID, false).Scan(&refreshedJob).Error
```

Comment:
Is it possible to not use `SKIP_LOCKED` in the query? The reason is that `worker` become unable to identify whether the row already has a lock or really doesn't exists.

Reply:
there will be no case the row will be empty because the record is not there in DB because all job already persist to DB before. I think case that I havent covered is when query return err because of issue with DB. I think for this case we can requeue the job after delay for several seconds

Reply:
Then what is this code for?
```
	if err == gorm.ErrRecordNotFound {
		tx.Rollback()
		log.Warnf("Job records not found with id: %d", job.ID)
		return
	}
```

Reply:
to return early, this indicate that job already success or locked in other process. The logging message should be improve, I'll change this